[Nicholas Arlin Halim] iP

daytona65 opened the pull request:
# DukeMeister3000
> "Tasks not complete, until DukeMeister3000 I used" - Mister Yoda

DukeMeister3000 tracks your tasks and gives you that edge to complete them on time. It is
- extremely intuitive
- fast
- ~~quite~~ ~~very~~ ABSOLUTELY simple to use

Simply
1. Download it [here](https://github.com/daytona65/ip/releases/tag/v0.2-fix)
2. Run it as a jar file
3. Start typing your tasks
4. And there you have it!

Let DukeMeister3000 manage your tasks for you! 💯 

Some features:
- [ ] Deadline setting
- [ ] Display your current tasks
- [ ] Stores your tasks when you stop the program and accesses them again on the next startup

If you are a Java programmer, you can review the code easily with its well designed code and JavaDocs. Here's the `newEvent` method:
```
    /**
     * Creates a new Event task and prints a message for the user
     * confirming the addition of the task to the Task List.
     *
     * @param input The user input.
     * @param taskSize The current Task List size.
     * @return The created task.
     */
    public Task newEvent(String input, int taskSize) throws DukeException {
        int fromIndex = input.indexOf(" /from ");
        int toIndex = input.indexOf(" /to ");
        Task task = new Event(input.substring(6, fromIndex), input.substring(fromIndex + 7, toIndex),
                input.substring(toIndex + 5));
        Ui.addMessage(task, taskSize);
        return task;
    }
```

Commit:
Add Gradle support

Commit:
Level 1

Commit:
Level 2

Commit:
Level 3

Commit:
Level 4

Commit:
Level 5

Commit:
Fixed Level 5 and added text ui testing increment

Commit:
Level-6

Dangabit commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,137 @@
+
 public class Duke {
+    private static Scanner sc;
+    private static List<Task> list;
```
Perhaps a better variable name than list here to indicate what the list holds?
```suggestion
    private static List<Task> tasks;
```

Dangabit commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,13 @@
+public class Deadline extends Task {
+    protected String by;
```
I think a more descriptive variable name would be more helpful.
```suggestion
    protected String dueDate;
```

Dangabit commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,137 @@
+            } else if (isMark(input, list.size())) {
+                int taskindex = Integer.parseInt(input.substring(5)) - 1;
+                list.get(taskindex).markAsDone();
+                System.out.println("Solid! I'm marking this task as done:\n" + list.get(taskindex).toString());
```
Line may have been too long, perhaps separate into multiple lines to be more aligned with the coding standard. There are also more similar issues further below.

Dangabit commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,13 @@
+public class Deadline extends Task {
```
I believe organising your Deadline, Todo, Event and task into a single package will help with the structure of your files. Also helps as you implement more classes.

Dangabit commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,137 @@
+        }
+    }
+
+    public static boolean isMark(String input, int listSize) {
```
All your methods are aptly named with the right names and format. Nice!

Dangabit reviewed (commented):
Coding standard wise it mostly look fine!
Some possible improvements are commented.

Commit:
Add save function

KSunil2001 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,137 @@
+                break;
+            } else if ("list".equals(input)) {
+                System.out.println("Here are the tasks in your list:");
+                for (int i = 0; i < list.size(); i++) {
+                    int j = i + 1;
+                    System.out.println(j + "." + list.get(i).toString());
+                }
+            } else if (isMark(input, list.size())) {
+                int taskindex = Integer.parseInt(input.substring(5)) - 1;
+                list.get(taskindex).markAsDone();
+                System.out.println("Solid! I'm marking this task as done:\n" + list.get(taskindex).toString());
+            } else if (isUnMark(input, list.size())) {
+                int taskindex = Integer.parseInt(input.substring(7)) - 1;
+                list.get(taskindex).markAsNotDone();
+                System.out.println("Aight, marking this as not done:\n" + list.get(taskindex).toString());
+            } else if (isDelete(input, list.size())) {
+                int taskindex = Integer.parseInt(input.substring(7)) - 1;
+                String removed = list.get(taskindex).toString();
+                list.remove(taskindex);
+                System.out.println("Swee! One less task to go! Removing...\n" + removed);
+            } else { //Task Creation
+                Task task = null;
+                if (isToDo(input)) {
+                    task = new ToDo(input);
+                    list.add(task);
+                } else if (isDeadline(input)) {
+                    int index = input.indexOf(" /by ");
+                    task = new Deadline(input.substring(0, index), input.substring(index + 5));
+                    list.add(task);
+                } else if (isEvent(input)) {
+                    int fromdex = input.indexOf(" /from ");
+                    int todex = input.indexOf(" /to ");
+                    task = new Event(input.substring(0, fromdex), input.substring(fromdex + 7, todex), input.substring(todex + 5));
+                    list.add(task);
+                } else {
+                    throw new DukeException("What are you saying?\n" + "Please input a task with either todo, deadline or event prefixed!");
+                }
+                System.out.println("Roger. This task has been added:\n" + "  " + task.toString());
+                System.out.println("Now you have " + list.size() + " tasks in your list.");
+            }
```
Maybe you could use switch case statements instead of if-else statements for better readability? 🤔

KSunil2001 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,137 @@
+
+    public static boolean isDeadline(String input) throws DukeException {
+        if (input.length() >= 8 && input.startsWith("deadline")) {
+            if (input.equals("deadline") || input.substring(8).isBlank() || input.equals("deadline /by") || input.equals("deadline /by ")) {
```
I would recommend breaking this statement into 2 lines to make the code look more neat.

KSunil2001 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,137 @@
+            if (input.equals("event") || input.substring(5).isBlank()) {
+                throw new DukeException("EVENT needs a description!");
+            }
+                int fromdex = input.indexOf(" /from ");
+                int todex = input.indexOf(" /to ");
+                if (fromdex + 7 > todex) {
+                    throw new DukeException("What are you saying?");
+                }
```
There seems to be additional indentation to these 5 lines of code.

KSunil2001 commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,37 @@
+
+    @Override
+    public String toString() {
+        return "[" + getStatusIcon() + "] " + getDescription();
```
Is getDescription() necessary here? 🤔 
```suggestion
        return "[" + getStatusIcon() + "] " + description;
```

KSunil2001 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,137 @@
+            } else return input.startsWith("todo ");
+        }
+        return false;
+    }
```
I like how you have defined your functions clearly and also applied the KISS rule. Well done 😄

KSunil2001 reviewed (commented):
LGTM, almost. Just a few nits to fix. 
Overall, the code is well done 😄

Commit:
Add Date and Time function

Commit:
Extract more classes for OOP

Commit:
Add duke, task and exception packages

Commit:
Merge branch 'add-gradle-support'

Commit:
Add JUnit tests for the 3 task classes

Commit:
Create new release

Commit:
Add JavaDoc comments and fixed coding standards

Commit:
Add coding standards in JavaDoc branch

Commit:
Add find command

Commit:
Merge tag 'A-CodingStandard'

Commit:
Merge branches JavaDoc, CodingStandard and Level-9

Commit:
Fix description bugs in all three tasks

Commit:
Fix Gradle

Commit:
GUI up to Dialog Box

Commit:
Clip image to circle and fix paddings

Commit:
Link GUI to Duke

Commit:
Add command package

Commit:
Command classes: extract a parent class Command

Command classes are all user commands parsed from user inputs, and
are executable. Extracting a Command class allows for better OOP
and modularity of the code.

Execution of commands is extracted from Parser class into Command
classes. Each command is a subclass of the abstract Command class
and each command implements execute() differently.

This allows for modularity and better OOP so more commands can easily
be added in the future. The code is also easier to understand and
functionality and execution of the commands are abstracted away
into the corresponding Command classes instead of just the Parser class.

Commit:
Fix JUnit tests

Commit:
Add assertions to Command, Task and Duke

Commit:
Improve code quality

Commit:
Level-10 tag

Commit:
Add more assertions

Commit:
Improve code quality

Commit:
Change some of Duke's responses

Commit:
Change comments in GUI

Commit:
Add @Override to execute() of subclasses of Command

daytona65 committed:
Merge pull request #2 from daytona65/branch-A-Assertions

Add assertions in Command, Task and Duke

Commit:
Merge branch 'master' into branch-A-CodeQuality

daytona65 committed:
Merge pull request #3 from daytona65/branch-A-CodeQuality

Improve code quality

Commit:
Merge branch 'branch-Level-10'

Commit:
Merge tag 'branch-A-CodeQuality'

Commit:
Merge branch 'master' of https://github.com/daytona65/ip

Commit:
Change product name to DukeMeister3000

Commit:
Implement Priority enum, PriorityCommand and all of its corresponding functions

Commit:
Add Priority Enums

Commit:
Make changes to GUI to make it look better, such as circle cropping user and Gigachad icons

Commit:
Add more personalised phrases for Duke

Commit:
Add more handling for edge cases

Commit:
Merge branch 'branch-BCD-Extension'

Commit:
Merge branch 'branch-A-BetterGui'

Commit:
Merge branch 'branch-A-Personality'

Commit:
Merge branch 'branch-A-MoreErrorHandling'

Commit:
Update build

Commit:
Add Ui.png to /docs

Commit:
Add user guide

Commit:
Update User Guide

Commit:
Fix pget indexing to 1 indexed commands

Commit:
Update UG

daytona65 committed:
Merge pull request #4 from daytona65/UG-update

Update UG

Commit:
Gigachad

daytona65 committed:
Merge pull request #5 from daytona65/UG-update

Gigachad

Commit:
Add spacings to UG

Commit:
Merge branch 'master' of https://github.com/daytona65/ip

Commit:
Add space UG
